### PR TITLE
Remove `Monad` and `Applicative` constraints on the `Parallel` class

### LINKED ISSUE #24: Is it possible to relax constraint from `Monad` to `Applicative` in `Parallel` class?
[I have an `Applicative`](https://github.com/paluh/purescript-polyform/blob/master/src/Polyform/Validator.purs#L21) (which is not a `Monad`) and [I have another `Applicative`](https://github.com/paluh/purescript-polyform/blob/master/src/Polyform/Validator/Par.purs#L15) which with additional constraints provides parallel apply for my initial type. It seems that this is the purpose of `Parallel` class - to define such a relation between types.
If I understand it correctly `Parallel` requires me to provide `Monad` instance for my "non parallel type". Why is this the case? Wouldn't it be enough to just require an `Applicative` instance?

### PR DESCRIPTION
**Pull request by artemisSystem**
**Description of the change**

Both types now only require `Apply`. This is to allow more types to have a `Parallel` instance.

Closes #24 

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [x] Linked any existing issues or proposals that this pull request should close
- [x] Updated or added relevant documentation
- [N/A] Added a test for the contribution (if applicable)

### CODE REVIEW

File: src/Control/Parallel/Class.purs [hunk 1]
```diff
@@ -19,13 +19,15 @@ import Data.Profunctor.Star (Star(..))
...
+-- | The `Parallel` class abstracts over pairs of `Apply`s where one of them
+-- | (`m`) composes sequentially, and the other (`f`) composes in parallel.
+-- | `m` is usually a `Monad`, which enforces the sequential nature of its
+-- | composition, but it doesn't need to be.
+class (Apply m, Apply f) <= Parallel f m | m -> f, f -> m where
   parallel :: m ~> f
   sequential :: f ~> m
 
-instance monadParExceptT :: Parallel f m => Parallel (Compose f (Either e)) (ExceptT e m) where
+instance monadParExceptT :: (Parallel f m, Apply (ExceptT e m)) => Parallel (Compose f (Either e)) (ExceptT e m) where
```
Comment (garyb) [inline src/Control/Parallel/Class.purs:30]: These constraint changes don't seem right, `ExceptT e m` is already `Apply` if `m` is `Monad`: https://github.com/purescript/purescript-transformers/blob/be72ab52357d9a665cbf93d73ba1c07c4b0957ee/src/Control/Monad/Except/Trans.purs#L58

Same goes for `MaybeT`.

Comment (artemisSystem) [reply src/Control/Parallel/Class.purs:30]: But m isn't a Monad anymore, that's the point of the pr, and that's why i had to add these constraints

Comment (garyb) [reply src/Control/Parallel/Class.purs:30]: Oh sorry, I misread the diff. In that case, the constraint here should be `Monad m`. It's impossible for there to be an `Apply (ExceptT e m)` instance otherwise anyway.

Comment (garyb) [reply src/Control/Parallel/Class.purs:30]: Do you have an example of a case where this new formulation is useful, by the way?

Comment (artemisSystem) [reply src/Control/Parallel/Class.purs:30]: > the constraint here should be Monad m. It's impossible for there to be an Apply (ExceptT e m) instance otherwise anyway.

That's a good point, i didn't think of that. I thought `Monad m` would be more restrictive, but i see how in this function, it isn't. I agree that `Monad m` is clearer and better, in that case.

> Do you have an example of a case where this new formulation is useful, by the way?

My use case is that i want to try making a parser that is applicative, but not a monad, so that its structure can be observed without actually running it on something (and it'll be built up from a symbolic representation, kinda like explained towards the end of this post: https://cronokirby.com/posts/2020/10/lexical-combinators/).

Then i thought i could have a newtype around that, which instead composes parsers in a way that lets them parse in any order (for example, parsing command line arguments). This would be a good candidate for a `Parallel` instance, i thought. You'd be able to use `parApply` (or [`Par.ado`](https://pursuit.purescript.org/packages/purescript-qualified-do/2.2.0/docs/QualifiedDo.ParApply)) to make permutable sub-parsers in an overall sequential parser.

Basically i think the "parse these applicatively combined parsers in any order" is a good parallel analogy to "parse these applicatively combined parsers in the order they were combined", and there are reasons to make sequential parsers that are not monads.

**Comment by MonoidMusician** [comment]
I think this looks good.

I wonder if a compromise would to still require `Applicative`. It's pretty rare that I run into an `Apply` that isn't `Applicative` (unless it is specifically constructed that way like a free construction, or is `Map k`). I think that would mean less breaking changes for most consumers (see `parTraverse` and friends).

But I would guess most people are probably using `Parallel` for `Aff` (or other concrete types), so they won't notice either way.

**Comment by artemisSystem** [comment]
To provide more context for the motivation, say you have a parser applicative (not monad, because being only applicative lets it be statically analyzed, i.e. you can interpret a free applicative while ignoring the result type of every intermediate calculation, something you can not do with a free monad).

the parser `token A *> token B *> token C` will only match `"ABC"`.
we could have a parallel version of this parser: `parallel (token A) *> parallel (token B) *> parallel (token C)`, which we could use to parse strings like `"CBA"`, `"ACB"`, etc.

it would be useful to give this a `Parallel` instance, which is currently not possible because the sequential parser is not a monad and `Parallel` requires that

**Comment by artemisSystem** [comment]
> I wonder if a compromise would to still require Applicative.

The potential breaking changes would only be for functions that are general over `Parallel` and implemented using `parTraverse`/friends, right? I think that would be a very small breakage in that case, if any.

The reason i went for `Apply` is that it's the least restrictive but still useful constraint. And i don't see a big reason why this shouldn't be as general/permissive as possible

**Review by MonoidMusician — APPROVED** [review]

**Review by garyb — APPROVED** [review]
